Automattic/jetpack
# Post by Email: Refactoring and Unit Tests

#### Description
_Post by Email_ module has two UI widgets:
- _"Post by email"_ section on the "_Jetpack Settings -> Writing"_ page
- _"Post by Email"_ section in the user profile
These two interfaces utilize different API endpoints, and that leads to duplicated code.

The goal of this PR is to minimize the code duplication, deprecate the older WP-AJAX endpoints in favor of the REST API, improve the test coverage of the module, and refactor the module to improve the code quality.

Nonetheless, the PR does not modify the module behavior in any way, and all deprecated functionality is still available (although it does throw "doing it wrong" warnings).

No issue was created.

#### Changes proposed in this Pull Request:
* Deprecating the WP-AJAX endpoints (`jetpack_post_by_email_enable`, `jetpack_post_by_email_regenerate`, `jetpack_post_by_email_disable`).
* Switching the "User Profile -> Post by Email" UI from WP-AJAX to REST API.
* Rewriting the jQuery-based JavaScript to VanillaJS.
* Moving REST requests proxying from `Jetpack_Core_API_Data` to the `post-by-email` module (as suggested by the @todo tag).
* Adding integration tests for the module's REST API.

#### Testing instructions:
0. Run unit tests (`yarn docker:phpunit`).
1. Load the _"Jetpack Settings -> Writing"_ and scroll down to the _"Post by email"_ section: /wp-admin/admin.php?page=jetpack#/writing
2. Enable module. If the module wasn't previously activated and the email field is empty, click on the _"Create address"_ button. **The email address should appear.**
3. Disable module. **The email field should be grayed out and disabled.**
4. Enable module again. The email address should show up in the input field, and _"Regenerate address"_ button will appear. Click on that button and **make sure that email address is replaced with a new one.**
<img width="1057" alt="post-by-email-1" src="https://user-images.githubusercontent.com/1341249/77713006-4e2ac680-6fa3-11ea-9043-3d443f3caf28.png">

5. Open the _"User Profile"_ page and scroll down to the _"Post by Email"_ section: /wp-admin/profile.php
6. **Make sure the email in the form matches the one on step 4.**
7. Click _"Regenerate Address"_. **The email address should be replaced with a new one.**
8. Click _"Disable Post By Email"_. **The form should be replaced with the _"Enable Post By Email"_ button.**
9. Click _"Enable Post By Email". **The form should reappear.**
<img width="706" alt="post-by-email2" src="https://user-images.githubusercontent.com/1341249/77713345-26882e00-6fa4-11ea-94d9-d26ba0b3443e.png">


#### Proposed changelog entry for your changes:
* Post by Email module: `WP-AJAX` endpoints are deprecated in favor of the REST API.
